Add migration from WF to Quarkus guide (#275)
keycloak/keycloak-web: guides/migration/migrating-to-quarkus.adoc

:guide-title: Migrating to Quarkus distribution
:guide-summary: Migrate to the new Quarkus distribution from the legacy WildFly distribution

In Keycloak 17 the default distribution is now powered by Quarkus, while the legacy WildFly powered distribution will still be around until June 2022 we highly recommend starting the migration as soon as possible.

The new distribution introduces a number of breaking changes, including:

* Configuring Keycloak has significantly changed
* Quarkus is not an application server, but rather a framework to build applications
* Custom providers are packaged and deployed differently

Before undertaking the migration we highly recommend reading through the new https://www.keycloak.org/guides#server[Server Guides] to understand how to install and configure the new distribution.

== Migrating configuration

The WildFly distribution of Keycloak used complicated XML files for configuration, resulting in the need for a CLI tool (`jboss-cli`) to manipulate these files. These files also brought complication to upgrading, where an error-prone script was used to upgrade the config from a previous version.

> The new Quarkus powered distribution leverages a simple configuration file instead, with corresponding CLI arguments and environment variables as options, making it significantly easier to configure Keycloak. However, this results in the ability to automatically migrate the configuration from the previous distribution.

Reviewer: The sentence "However, this results in the ability to automatically migrate the configuration from the previous distribution." could be misunderstood imo, saying that you can automatically migrate wildfly config to quarkus. perhaps better:

"However, this results in the ability to automatically migrate the configuration from the previous quarkus distribution in the future."

or is here a "not" missing?

Author: Yeah, that's completely wrong. See the suggestion I have below. Does that work?

Reviewer: yep, works for me, thanks

> The new Quarkus powered distribution leverages a simple configuration file instead, with corresponding CLI arguments and environment variables as options, making it significantly easier to configure Keycloak. However, this results in the ability to automatically migrate the configuration from the previous distribution.

Author: ```suggestion
The new Quarkus powered distribution leverages a simple configuration file instead, with corresponding CLI arguments and environment variables as options, making it significantly easier to configure Keycloak.

Since the WildFly distribution was too flexible in how much it could be configured and tuned, as well as the approach for configuration changing so much, it is not possible to automatically migrate the configuration from the old WildFly distribution to the new Quarkus distribution.
```

To migrate to the new Quarkus powered distribution the first step is to understand what configuration changes you are applying to the legacy distribution, and apply those that are necessary to the new distribution by following the new https://www.keycloak.org/guides#server[Server Guides].

> One thing to note is the new distribution is a lot more opinionated when it comes to configuration. It aims to provide better defaults, with the need to configure less yourself. However, we may not always have the balance right, and there may be use-cases not covered.

Reviewer: ```suggestion
One thing to note is the new distribution is a lot more opinionated when it comes to configuration. It aims to provide better defaults, with the need to configure less yourself. However, we may not always have the balance right and there may be use-cases not covered, yet.
```

Author: See below for perhaps a better suggestion than to add `yet` to the end.

> One thing to note is the new distribution is a lot more opinionated when it comes to configuration. It aims to provide better defaults, with the need to configure less yourself. However, we may not always have the balance right, and there may be use-cases not covered.

Author: ```suggestion
One thing to note is the new distribution is a lot more opinionated when it comes to configuration. It aims to provide better defaults, with the need to configure less yourself. However, we may not always have the balance right, and there may be use-cases that are not currently covered.
```

If you are unable to configure something that you need to tweak in the new distribution, please open a discussion in https://github.com/keycloak/keycloak/discussions/categories/keycloak-x-quarkus-distribution[GitHub Discussions].

Until a new release is available it is possible to configure the new distribution by directly applying Quarkus level configuration through the `conf/quarkus.properties` file. We recommend you use this sparingly as you will be applying configuration untested and unsupported by the Keycloak team.

## Quarkus is not an application server

Unlike WildFly, Quarkus is not an application server. While an application server can dynamically deploy applications, and alter what is loaded into memory at runtime, this is not possible on Quarkus.

Quarkus on the other hand brings immutability to containers, faster startup, and more predictability.

> While with the WildFly distribution you could hot-deploy custom providers, change the database vendor as a runtime configuration this is no longer supported.

Reviewer: ```suggestion
While with the WildFly distribution you could hot-deploy custom providers or change the database vendor as a runtime configuration, this is no longer supported.
```

Instead, the Quarkus distribution provides a separate build step that optimises the runtime. One important thing to note here is that the build step does not actually build the Keycloak sources, but rather just optimises the runtime through an augmentation process, which is fairly fast and able to fully optimise what is loaded into the runtime.

We recommend that you do this build step as a part of installing Keycloak, through CI, or by creating a custom container image that extends the base Keycloak image.

However, there is also an auto build mode that makes Keycloak behave more or less the same as the WildFly distribution in this regard. This comes with a startup time penalty, but is still able to optimise the runtime better than the WildFly distribution could.

## Migrating custom providers

Similarly to the WildFly distribution custom providers are deployed to Keycloak by copying them to a deployment directory. In the new distribution you should copy your providers to the `providers` directory instead of `standalone/deployments`, which no longer exists. Additional dependencies are also copied to the `providers` directory.

> With the new distribution there is no longer a separate classpath for custom providers, so you may need to be more careful with what additional dependencies you include. In addition, the `EAR` packaging format, and `jboss-deployment-structure.xml` files, is no longer supported.

Reviewer: ```suggestion
With the new distribution there is no longer a separate classpath for custom providers, so you may need to be more careful with what additional dependencies you include. In addition, the `EAR` packaging format, and `jboss-deployment-structure.xml` files, are no longer supported.
```
